src/shard.rs

```diff
@@ -332,6 +332,32 @@ impl<
         })
     }
 
+    pub fn retain<F>(&mut self, f: F)
+    where
+        F: Fn(&Key, &Val) -> bool,
+    {
+        let retained_tokens = self
+            .map
+            .iter()
+            .filter_map(|&idx| match self.entries.get(idx) {
+                Some((entry, _idx)) => match entry {
+                    Entry::Resident(r) => {
+                        if f(&r.key, &r.value) {
+                            let hash = self.hash(&r.key);
+                            Some((idx, hash))
+                        } else {
+                            None
+                        }
+                    }
+                    Entry::Placeholder(_) | Entry::Ghost(_) => None,
+                },
+                None => None,
+            });
+        for (idx, hash) in retained_tokens {
+            self.remove_internal(hash, idx);
```

Comment:
There is a compile error here, I have no idea how to fix it ...

<img width="669" alt="image" src="https://github.com/user-attachments/assets/547bb71d-4043-42cd-ba1e-96991579012a" />

Reply:
maybe just use `unsafe` to bypass the  mutable reference issue

Reply:
Update: I fixed this compile error without `unsafe`.